File: backend/app/schemas/project.py
from pydantic import BaseModel, field_validator, ConfigDict
from datetime import datetime
from typing import Optional, List, Union
from app.models.project import ProjectStatus
import json


class ProjectBase(BaseModel):
    title: str
    author: str
    genre: Union[str, List[str]]  # 支持字符串或字符串列表
    summary: Optional[str] = None
    target_readers: Optional[str] = None
    default_pov: Optional[str] = None
    style: Optional[str] = None
    target_words_per_chapter: int = 2000

    @field_validator('title', 'author')
    @classmethod
    def not_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError('不能为空')
        return v.strip()

    @field_validator('genre', mode='before')
    @classmethod
    def validate_genre(cls, v: Union[str, List[str], None]) -> str:
        # 处理 None 值
        if v is None:
            raise ValueError('类型不能为空')

        # 如果是列表，转换为 JSON 字符串
        if isinstance(v, list):
            if not v or len(v) == 0:
                raise ValueError('类型不能为空')
            # 去重并过滤空值
            genres = [g.strip() for g in v if g and g.strip()]
            if not genres:
                raise ValueError('类型不能为空')
            return json.dumps(list(set(genres)), ensure_ascii=False)

        # 如果已经是字符串，直接返回（前端会负责解析）
        if isinstance(v, str):
            if not v.strip():
                raise ValueError('类型不能为空')
            return v

        raise ValueError('类型格式不正确')


class ProjectCreate(ProjectBase):
    tags: Optional[str] = None
    background_template: Optional[str] = None


class ProjectUpdate(BaseModel):
    title: Optional[str] = None
    author: Optional[str] = None
    genre: Optional[str] = None
    tags: Optional[str] = None
    summary: Optional[str] = None
    target_readers: Optional[str] = None
    status: Optional[ProjectStatus] = None
    default_pov: Optional[str] = None
    style: Optional[str] = None
    style_keywords: Optional[str] = None
    language_style: Optional[str] = None
    sensory_focus: Optional[str] = None
    style_intensity: Optional[int] = None
    target_words_per_chapter: Optional[int] = None
    background_template: Optional[str] = None


class ProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    title: str
    author: str
    genre: str  # 保持为字符串（JSON 格式）
    summary: Optional[str] = None
    target_readers: Optional[str] = None
    status: str  # 改为字符串，避免枚举验证问题
    default_pov: Optional[str] = None
    style: Optional[str] = None
    target_words_per_chapter: int
    tags: Optional[str] = None
    background_template: Optional[str] = None
    total_words: int
    total_chapters: int
    completion_rate: int
    created_at: datetime
    updated_at: datetime


class ProjectListResponse(BaseModel):
    projects: List[ProjectResponse]
    total: int

File: backend-java/src/main/java/com/novel/ai/model/entity/Project.java
package com.novel.ai.model.entity;

import java.time.LocalDateTime;

public class Project {
    private Integer id;
    private Integer userId;
    private String title;
    private String author;
    private String genre;
    private String tags;
    private String summary;
    private String targetReaders;
    private String status;
    private String defaultPov;
    private String style;
    private String styleKeywords;
    private String languageStyle;
    private String sensoryFocus;
    private Integer styleIntensity;
    private Integer targetWordsPerChapter;
    private String backgroundTemplate;
    private Integer totalWords;
    private Integer totalChapters;
    private Integer completionRate;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    // Constructors
    public Project() {}

    public Project(Integer userId, String title, String author) {
        this.userId = userId;
        this.title = title;
        this.author = author;
        this.status = "draft";
        this.styleIntensity = 70;
        this.targetWordsPerChapter = 2000;
        this.totalWords = 0;
        this.totalChapters = 0;
        this.completionRate = 0;
    }

    // Getters and Setters
    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public Integer getUserId() {
        return userId;
    }

    public void setUserId(Integer userId) {
        this.userId = userId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getGenre() {
        return genre;
    }

    public void setGenre(String genre) {
        this.genre = genre;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    public String getSummary() {
        return summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public String getTargetReaders() {
        return targetReaders;
    }

    public void setTargetReaders(String targetReaders) {
        this.targetReaders = targetReaders;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getDefaultPov() {
        return defaultPov;
    }

    public void setDefaultPov(String defaultPov) {
        this.defaultPov = defaultPov;
    }

    public String getStyle() {
        return style;
    }

    public void setStyle(String style) {
        this.style = style;
    }

    public String getStyleKeywords() {
        return styleKeywords;
    }

    public void setStyleKeywords(String styleKeywords) {
        this.styleKeywords = styleKeywords;
    }

    public String getLanguageStyle() {
        return languageStyle;
    }

    public void setLanguageStyle(String languageStyle) {
        this.languageStyle = languageStyle;
    }

    public String getSensoryFocus() {
        return sensoryFocus;
    }

    public void setSensoryFocus(String sensoryFocus) {
        this.sensoryFocus = sensoryFocus;
    }

    public Integer getStyleIntensity() {
        return styleIntensity;
    }

    public void setStyleIntensity(Integer styleIntensity) {
        this.styleIntensity = styleIntensity;
    }

    public Integer getTargetWordsPerChapter() {
        return targetWordsPerChapter;
    }

    public void setTargetWordsPerChapter(Integer targetWordsPerChapter) {
        this.targetWordsPerChapter = targetWordsPerChapter;
    }

    public String getBackgroundTemplate() {
        return backgroundTemplate;
    }

    public void setBackgroundTemplate(String backgroundTemplate) {
        this.backgroundTemplate = backgroundTemplate;
    }

    public Integer getTotalWords() {
        return totalWords;
    }

    public void setTotalWords(Integer totalWords) {
        this.totalWords = totalWords;
    }

    public Integer getTotalChapters() {
        return totalChapters;
    }

    public void setTotalChapters(Integer totalChapters) {
        this.totalChapters = totalChapters;
    }

    public Integer getCompletionRate() {
        return completionRate;
    }

    public void setCompletionRate(Integer completionRate) {
        this.completionRate = completionRate;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }
}